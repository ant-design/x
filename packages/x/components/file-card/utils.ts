export function getSize(size: number) {
  let retSize = size;
  const units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB'];
  let unitIndex = 0;

  while (retSize >= 1024 && unitIndex < units.length - 1) {
    retSize /= 1024;
    unitIndex++;
  }

  return `${retSize.toFixed(0)} ${units[unitIndex]}`;
}

export function matchExt(suffix: string, ext: string[]) {
  return ext.some((e) => suffix.toLowerCase() === `.${e}`);
}
