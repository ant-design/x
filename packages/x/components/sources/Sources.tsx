import { RightOutlined } from '@ant-design/icons';
import { Popover } from 'antd';
import classnames from 'classnames';
import type { CSSMotionProps } from 'rc-motion';
import CSSMotion from 'rc-motion';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import pickAttrs from 'rc-util/lib/pickAttrs';
import React from 'react';
import useProxyImperativeHandle from '../_util/hooks/use-proxy-imperative-handle';
import useXComponentConfig from '../_util/hooks/use-x-component-config';
import initCollapseMotion from '../_util/motion';
import { useXProviderContext } from '../x-provider';
import CarouselCard from './components/CarouselCard';
import useStyle from './style';

export type SemanticType = 'root' | 'title' | 'content';

export interface SourcesItem {
  key?: React.Key;
  title: React.ReactNode;
  url?: string;
  icon?: React.ReactNode;
  description?: React.ReactNode;
}

export interface SourcesProps
  extends Omit<React.HTMLAttributes<HTMLDivElement>, 'title' | 'onClick'> {
  prefixCls?: string;
  style?: React.CSSProperties;
  styles?: Partial<Record<SemanticType, React.CSSProperties>>;
  className?: string;
  classNames?: Partial<Record<SemanticType, string>>;
  rootClassName?: string;
  title?: React.ReactNode;
  expandIconPosition?: 'start' | 'end';
  items?: Array<SourcesItem>;
  defaultExpanded?: boolean;
  expanded?: boolean;
  onExpand?: (expand: boolean) => void;
  onClick?: (item: SourcesItem) => void;
  inline?: boolean;
  popoverOverlayWidth?: number | string;
}

type SourcesRef = {
  nativeElement: HTMLElement;
};

const Sources: React.ForwardRefRenderFunction<SourcesRef, SourcesProps> = (props, ref) => {
  const {
    prefixCls: customizePrefixCls,
    style,
    styles = {},
    className,
    rootClassName,
    classNames = {},
    title,
    expandIconPosition = 'start',
    items,
    defaultExpanded = true,
    expanded,
    onExpand,
    children,
    inline = false,
    popoverOverlayWidth = 300,
    ...restProps
  } = props;

  // ============================ Prefix ============================

  const { direction, getPrefixCls } = useXProviderContext();
  const prefixCls = getPrefixCls('sources', customizePrefixCls);
  const [hashId, cssVarCls] = useStyle(prefixCls);

  // ======================= Component Config =======================

  const contextConfig = useXComponentConfig('sources');
  const domProps = pickAttrs(restProps, {
    attr: true,
    aria: true,
    data: true,
  });

  // ============================= Refs =============================
  const sourcesRef = React.useRef<HTMLDivElement>(null);
  useProxyImperativeHandle(ref, () => {
    return {
      nativeElement: sourcesRef.current!,
    };
  });

  const mergedCls = classnames(
    prefixCls,
    contextConfig.className,
    className,
    rootClassName,
    classNames.root,
    hashId,
    cssVarCls,
    {
      [`${prefixCls}-inline`]: inline,
      [`${prefixCls}-rtl`]: direction === 'rtl',
    },
  );

  // ============================  Collapsible ============================

  const [isExpand, setIsExpand] = useMergedState(defaultExpanded, {
    value: expanded,
    onChange: onExpand,
  });

  const collapseMotion: CSSMotionProps = {
    ...initCollapseMotion(),
    motionAppear: false,
    leavedClassName: `${prefixCls}-content-hidden`,
  };

  const ContentNode = items ? (
    <ul className={`${prefixCls}-list`}>
      {items.map((item, index) => (
        <li
          key={item.key || index}
          className={`${prefixCls}-list-item`}
          onClick={() => props.onClick?.(item)}
        >
          <a
            className={`${prefixCls}-link`}
            href={item.url}
            target="_blank"
            rel="noopener noreferrer"
          >
            {item.icon && <span className={`${prefixCls}-link-icon`}>{item.icon}</span>}
            <span className={`${prefixCls}-link-title`}>{item.title}</span>
          </a>
        </li>
      ))}
    </ul>
  ) : (
    children
  );

  return (
    <div
      ref={sourcesRef}
      {...domProps}
      className={mergedCls}
      style={{
        ...contextConfig.style,
        ...contextConfig.styles.root,
        ...style,
        ...styles.root,
      }}
    >
      {inline ? (
        <Popover
          content={<CarouselCard prefixCls={prefixCls} items={items} onClick={props.onClick} />}
          open={inline ? undefined : false}
          styles={{ body: { width: popoverOverlayWidth } }}
          placement="top"
          forceRender
        >
          <div
            className={classnames(`${prefixCls}-title-wrapper`, classNames.title)}
            style={styles.title}
          >
            <span className={`${prefixCls}-title`}>{title}</span>
          </div>
        </Popover>
      ) : (
        <>
          <div
            className={classnames(
              `${prefixCls}-title-wrapper`,
              `${prefixCls}-icon-position-${expandIconPosition}`,
              classNames.title,
            )}
            onClick={() => setIsExpand(!isExpand)}
            style={styles.title}
          >
            <RightOutlined className={`${prefixCls}-title-down-icon`} rotate={isExpand ? 90 : 0} />
            <span className={`${prefixCls}-title`}>{title}</span>
          </div>
          <CSSMotion {...collapseMotion} visible={isExpand}>
            {({ className: motionClassName, style }, motionRef) => (
              <div
                className={classnames(`${prefixCls}-content`, motionClassName, classNames.content)}
                ref={motionRef}
                style={{ ...style, ...styles.content }}
              >
                {ContentNode}
              </div>
            )}
          </CSSMotion>
        </>
      )}
    </div>
  );
};

const ForwardSources = React.forwardRef(Sources);

if (process.env.NODE_ENV !== 'production') {
  ForwardSources.displayName = 'Sources';
}

export default ForwardSources;
