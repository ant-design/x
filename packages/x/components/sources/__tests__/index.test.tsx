import React from 'react';

import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { fireEvent, render, waitFakeTimer } from '../../../tests/utils';
import Sources from '../index';

const items = [
  {
    title: '1. Data source',
    url: 'https://x.ant.design/components/overview',
  },
  {
    title: '2. Data source',
    url: 'https://x.ant.design/components/overview',
  },
  {
    title: '3. Data source',
    url: 'https://x.ant.design/components/overview',
  },
];

describe('Sources Component', () => {
  mountTest(() => <Sources title={'Test'} items={items} />);

  rtlTest(() => <Sources title={'Test'} items={items} />);

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('Sources component work', () => {
    const { container } = render(<Sources title={'Test'} items={items} />);
    const element = container.querySelector<HTMLDivElement>('.ant-sources');
    expect(element).toBeTruthy();
    expect(element).toMatchSnapshot();
  });

  it('Sources should support custom style and styles', () => {
    const { container } = render(
      <Sources
        title="Test"
        items={items}
        style={{ backgroundColor: 'red' }}
        styles={{ root: { padding: '10px' } }}
      />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-sources');
    expect(element).toHaveStyle('background-color: red');
    expect(element).toHaveStyle('padding: 10px');
  });

  it('Sources should support custom className and classNames', () => {
    const { container } = render(
      <Sources
        title="Test"
        items={items}
        className="test-className"
        classNames={{ root: 'test-className2' }}
      />,
    );
    const element = container.querySelector<HTMLDivElement>('.ant-sources');
    expect(element).toHaveClass('test-className');
    expect(element).toHaveClass('test-className2');
  });

  it('Sources should support expandIconPosition', () => {
    const { container } = render(<Sources title="Test" items={items} expandIconPosition="end" />);
    const element = container.querySelector<HTMLDivElement>('.ant-sources-icon-position-end');
    expect(element).toBeTruthy();
  });

  it('Sources should support controlled expanded state', async () => {
    const { container } = render(<Sources title="Test" items={items} defaultExpanded={false} />);

    expect(container.querySelector('.ant-sources-content')).toBeNull();
    fireEvent.click(container.querySelector('.ant-sources-title-wrapper')!);
    await waitFakeTimer();
    expect(container.querySelector('.ant-sources-content')).toBeTruthy();
  });

  it('Sources should support inline mode', () => {
    const { container } = render(<Sources title="Test" items={items} inline />);
    const element = container.querySelector<HTMLDivElement>('.ant-sources-inline');
    expect(element).toBeTruthy();
  });

  it('Sources should support items with all properties', () => {
    const fullItems = [
      {
        key: 'test-key',
        title: 'Test Title',
        url: 'https://test.com',
        icon: <span data-testid="test-icon">icon</span>,
        description: 'Test description',
      },
    ];

    const { container, getByTestId } = render(<Sources title="Test" items={fullItems} />);

    const icon = getByTestId('test-icon');
    expect(icon).toBeTruthy();

    const link = container.querySelector<HTMLAnchorElement>('.ant-sources-link');
    expect(link).toHaveAttribute('href', 'https://test.com');
  });

  it('Sources should support children instead of items', () => {
    const { container } = render(
      <Sources title="Test">
        <div data-testid="custom-content">Custom content</div>
      </Sources>,
    );

    const customContent = container.querySelector('[data-testid="custom-content"]');
    expect(customContent).toBeTruthy();
  });

  it('Sources supports ref', () => {
    const ref = React.createRef<any>();
    render(<Sources ref={ref} title={'Test'} items={items} />);
    expect(ref.current).not.toBeNull();
  });
});
