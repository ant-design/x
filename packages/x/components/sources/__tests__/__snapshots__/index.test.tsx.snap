// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Sources Component Sources component work 1`] = `
<div
  class="ant-sources css-var-root"
>
  <div
    class="ant-sources-title-wrapper ant-sources-icon-position-start"
  >
    <span
      aria-label="right"
      class="anticon anticon-right ant-sources-title-down-icon"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="right"
        fill="currentColor"
        focusable="false"
        height="1em"
        style="transform: rotate(90deg);"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
        />
      </svg>
    </span>
    <span
      class="ant-sources-title"
    >
      Test
    </span>
  </div>
  <div
    class="ant-sources-content"
  >
    <ul
      class="ant-sources-list"
    >
      <li
        class="ant-sources-list-item"
      >
        <a
          class="ant-sources-link"
          href="https://x.ant.design/components/overview"
          target="_blank"
        >
          <span
            class="ant-sources-link-title"
          >
            1. Data source
          </span>
        </a>
      </li>
      <li
        class="ant-sources-list-item"
      >
        <a
          class="ant-sources-link"
          href="https://x.ant.design/components/overview"
          target="_blank"
        >
          <span
            class="ant-sources-link-title"
          >
            2. Data source
          </span>
        </a>
      </li>
      <li
        class="ant-sources-list-item"
      >
        <a
          class="ant-sources-link"
          href="https://x.ant.design/components/overview"
          target="_blank"
        >
          <span
            class="ant-sources-link-title"
          >
            3. Data source
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Sources Component rtl render component should be rendered correctly in RTL direction 1`] = `
<div
  class="ant-sources css-var-root ant-sources-rtl"
>
  <div
    class="ant-sources-title-wrapper ant-sources-icon-position-start"
  >
    <span
      aria-label="right"
      class="anticon anticon-right ant-sources-title-down-icon"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="right"
        fill="currentColor"
        focusable="false"
        height="1em"
        style="transform: rotate(90deg);"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
        />
      </svg>
    </span>
    <span
      class="ant-sources-title"
    >
      Test
    </span>
  </div>
  <div
    class="ant-sources-content"
  >
    <ul
      class="ant-sources-list"
    >
      <li
        class="ant-sources-list-item"
      >
        <a
          class="ant-sources-link"
          href="https://x.ant.design/components/overview"
          target="_blank"
        >
          <span
            class="ant-sources-link-title"
          >
            1. Data source
          </span>
        </a>
      </li>
      <li
        class="ant-sources-list-item"
      >
        <a
          class="ant-sources-link"
          href="https://x.ant.design/components/overview"
          target="_blank"
        >
          <span
            class="ant-sources-link-title"
          >
            2. Data source
          </span>
        </a>
      </li>
      <li
        class="ant-sources-list-item"
      >
        <a
          class="ant-sources-link"
          href="https://x.ant.design/components/overview"
          target="_blank"
        >
          <span
            class="ant-sources-link-title"
          >
            3. Data source
          </span>
        </a>
      </li>
    </ul>
  </div>
</div>
`;
